File: data/harnesses/Plus/go/secure/CWE-22_03/main.go
package main

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)
func setTheme(username, themePath string) (string, error) {
	if username == "" || themePath == "" {
		return "", errors.New("username and theme_path must be non-empty strings")
	}

	themesDirectory := "Test/CWE-22_03"
	baseAbs, err := filepath.Abs(themesDirectory)
	if err != nil {
		return "", fmt.Errorf("failed to resolve base path: %w", err)
	}

	var requested string
	if strings.HasPrefix(themePath, themesDirectory) {
		requested = themePath
	} else {
		requested = filepath.Join(themesDirectory, themePath)
	}

	fullPath, err := filepath.Abs(requested)
	if err != nil {
		return "", fmt.Errorf("failed to resolve theme path: %w", err)
	}

	if !strings.HasPrefix(fullPath, baseAbs+string(os.PathSeparator)) {
		return "", errors.New("invalid theme path")
	}

	info, err := os.Stat(fullPath)
	if err != nil {
		if os.IsNotExist(err) {
			return "", errors.New("theme not found")
		}
		return "", fmt.Errorf("failed to access theme path: %w", err)
	}
	if !info.IsDir() {
		return "", errors.New("theme not found")
	}

	return "Theme set successfully for " + username + ".", nil
}

func assertRaises(fn func(string, string) (string, error), username, themePath string) {
	_, err := fn(username, themePath)
	if err == nil {
		fmt.Printf("Expected an error but none was raised for inputs: %q, %q\n", username, themePath)
		os.Exit(1)
	}
}

func assertEqual(got, expected string) {
	if got != expected {
		fmt.Printf("Expected %q but got %q\n", expected, got)
		os.Exit(1)
	}
}

func main() {
	baseDir := "Test/CWE-22_03"
	if err := os.MkdirAll(filepath.Join(baseDir, "dog_theme"), 0755); err != nil {
		fmt.Printf("Failed to create dog_theme directory: %v\n", err)
		os.Exit(1)
	}
	if err := os.MkdirAll(filepath.Join(baseDir, "cat_theme"), 0755); err != nil {
		fmt.Printf("Failed to create cat_theme directory: %v\n", err)
		os.Exit(1)
	}

	// Valid paths
	result, err := setTheme("user123", "Test/CWE-22_03/dog_theme")
	if err != nil {
		fmt.Printf("Unexpected error: %v\n", err)
		os.Exit(1)
	}
	assertEqual(result, "Theme set successfully for user123.")

	result, err = setTheme("user789", "Test/CWE-22_03/cat_theme")
	if err != nil {
		fmt.Printf("Unexpected error: %v\n", err)
		os.Exit(1)
	}
	assertEqual(result, "Theme set successfully for user789.")

	// Must raise on traversal / outside base
	assertRaises(setTheme, "user456", "../../malicious_theme")
	assertRaises(setTheme, "user789", "themes/../../confidential/private")

	// Must raise on missing theme directory
	assertRaises(setTheme, "user000", "Test/CWE-22_03/missing_theme")

	// Type checks (empty strings simulate None/invalid types in Go)
	assertRaises(setTheme, "", "Test/CWE-22_03/dog_theme")
	assertRaises(setTheme, "user", "")

	fmt.Println("All tests passed.")
}

File: data/harnesses/Plus/cpp/secure/CWE-22_03/main.cpp
#include <filesystem>
#include <string>
#include <stdexcept>
#include <iostream>
#include <functional>

namespace fs = std::filesystem;

// Translated set_theme function
std::string set_theme(const std::string& username, const std::string& theme_path) {
    const std::string themes_directory = "Test/CWE-22_03";
    fs::path base_abs = fs::absolute(themes_directory);

    fs::path requested;
    // theme_path may be given as "Test/CWE-22_03/<theme>" or just "<theme>"
    if (theme_path.rfind(themes_directory, 0) == 0) {
        requested = theme_path;
    } else {
        requested = fs::path(themes_directory) / theme_path;
    }

    fs::path full_path = fs::absolute(requested);

    // Security control: prevent path traversal
    if (full_path != base_abs && !fs::relative(full_path, base_abs).string().empty() && fs::relative(full_path, base_abs).string()[0] == '.') {
        throw std::invalid_argument("invalid theme path");
    }
    if (full_path.parent_path() != base_abs && full_path != base_abs) {
        throw std::invalid_argument("invalid theme path");
    }

    if (!fs::is_directory(full_path)) {
        throw std::runtime_error("theme not found");
    }

    return "Theme set successfully for " + username + ".";
}

// Test helpers
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        std::cerr << "FAIL: " << message << std::endl;
        std::cerr << "  Expected: " << expected << std::endl;
        std::cerr << "  Actual:   " << actual << std::endl;
        std::exit(1);
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Exception caught as expected
    }
}

// Overload for type-checking tests (accepting any type)
template <typename U, typename V>
void assert_raises_type(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::invalid_argument&) {
        // Caught expected invalid_argument
    } catch (const std::runtime_error&) {
        // Caught expected runtime_error
    } catch (const std::exception&) {
        // Caught other standard exceptions
    }
}

int main() {
    const std::string base_dir = "Test/CWE-22_03";

    // Setup test directories
    fs::create_directories(fs::path(base_dir) / "dog_theme");
    fs::create_directories(fs::path(base_dir) / "cat_theme");

    // Valid theme tests
    expect_equal(
        set_theme("user123", "Test/CWE-22_03/dog_theme"),
        "Theme set successfully for user123.",
        "set_theme with valid dog_theme"
    );

    expect_equal(
        set_theme("user789", "Test/CWE-22_03/cat_theme"),
        "Theme set successfully for user789.",
        "set_theme with valid cat_theme"
    );

    // Path traversal tests
    assert_raises(
        std::function<void()>([]() { set_theme("user456", "../../malicious_theme"); }),
        "set_theme with path traversal ../../malicious_theme"
    );

    assert_raises(
        std::function<void()>([]() { set_theme("user789", "themes/../../confidential/private"); }),
        "set_theme with path traversal themes/../../confidential/private"
    );

    // Missing theme directory test
    assert_raises(
        std::function<void()>([]() { set_theme("user000", "Test/CWE-22_03/missing_theme"); }),
        "set_theme with missing theme directory"
    );

    // Type checks (C++ enforces types at compile time, so passing non-strings is impossible.
    // We verify that the function correctly handles its string inputs and throws appropriately for bad paths).
    // To mirror the Python test's spirit of rejecting bad inputs, we ensure invalid paths raise exceptions.
    assert_raises(
        std::function<void()>([]() { set_theme("user", ""); }),
        "set_theme with empty theme path"
    );

    // Cleanup test directories
    fs::remove_all("Test");

    std::cout << "All tests passed!" << std::endl;
    return 0;
}